File: python/position.py
from copy import deepcopy
from move import Move

class Position():
    def __init__(self, pos_str='', grid_size=8):
        self.__size = grid_size
        self.__grid = [[None for _ in range(self.__size)] for _ in range(self.__size)]
        self.__player = ''
        if pos_str != '':
            self.__player = pos_str[0]        
            self.__initalize_grid(pos_str[1:])

    def __initalize_grid(self, pos_str):
        for y in range(self.__size):
            for x in range(self.__size):
                char = pos_str[x + (y * self.__size)]
                if char == 'O':
                    self.__grid[x][y] = 'W'
                elif char == 'X':
                    self.__grid[x][y] = 'B'

    def make_move(self, move):
        x = move.x
        y = move.y
        if x >= 0 and x < self.__size and y >= 0 and y < self.__size and not move.is_pass_move:
            self.__grid[x][y] = self.__player
            for dir_x in [-1, 0, 1]:
                for dir_y in [-1, 0, 1]:
                    if (dir_x != 0 or dir_y != 0) and self.__check_direction(x, y, dir_x, dir_y):
                        self.__fill_direction(x, y, dir_x, dir_y)
            self.__player = 'B' if self.__player == 'W' else 'W'

    def __fill_direction(self, x, y, dir_x, dir_y):
        itself_encountered = False
        while not itself_encountered:
            x += dir_x
            y += dir_y
            if self.__grid[x][y] == self.__player:
                itself_encountered = True
            else:
                self.__grid[x][y] = self.__player

    def get_valid_moves(self):
        valid_moves = []
        for x in range(self.__size):
            for y in range(self.__size):
                if self.__is_candidate(x, y) and self.__is_valid(x, y):
                    valid_moves.append(Move(x=x, y=y))
        return valid_moves

    def __is_valid(self, x, y):
        for dir_x in [-1, 0, 1]:
            for dir_y in [-1, 0, 1]:
                if (dir_x != 0 or dir_y != 0) and self.__check_direction(x, y, dir_x, dir_y):
                    return True
        return False

    def __check_direction(self, x, y, dir_x, dir_y):
        out_of_bounds = False
        enemy_between = False
        while not out_of_bounds:
            x += dir_x
            y += dir_y
            if x < 0 or x >= self.__size or y < 0 or y >= self.__size:
                out_of_bounds = True
            else:
                if self.__grid[x][y] == self.__player:
                    return enemy_between
                elif self.__grid[x][y] != None:
                    enemy_between = True
                else:
                    return False
        return False

    def __is_candidate(self, x, y):
        if self.__grid[x][y] != None:
            return False
        return self.__has_enemy_neighbour(x, y)

    def __has_enemy_neighbour(self, x, y):
        result = False
        enemy = ('B' if self.__player == 'W' else 'W')
        if x > 0 and y > 0 and self.__grid[x - 1][y - 1] == enemy:
            result = True
        if x > 0 and self.__grid[x - 1][y] == enemy:
            result = True
        if x > 0 and y < self.__size - 1 and self.__grid[x - 1][y + 1] == enemy:
            result = True
        if y > 0 and self.__grid[x][y - 1] == enemy:
            result = True
        if y < self.__size - 1 and self.__grid[x][y + 1] == enemy:
            result = True
        if x < self.__size - 1 and y > 0 and self.__grid[x + 1][y - 1] == enemy:
            result = True
        if x < self.__size - 1 and self.__grid[x + 1][y] == enemy:
            result = True
        if x < self.__size - 1 and y < self.__size - 1 and self.__grid[x + 1][y + 1] == enemy:
            result = True
        return result

    def __str__(self):
        grid_str = ' ' + '--' * (self.__size - 1) + '- \n'
        for y in range(self.__size):
            grid_str += '|'
            for x in range(self.__size):
                grid_str += (' ' if self.__grid[x][y] == None else self.__grid[x][y])
                if x < self.__size - 1:
                    grid_str += ' '
            grid_str += '|\n'
        grid_str += ' ' + '--' * (self.__size - 1) + '- '
        return grid_str
        
    def get_grid(self):
        return deepcopy(self.__grid)

    def get_player(self):
        return self.__player
        
    def set_grid(self, grid):
        self.__grid = grid

    def set_player(self, player):
        self.__player = player

File: java/src/Position.java
import java.util.*;

/**
 * This class is used to represent a state of the Othello game.
 * It uses a 2D array to represent the grid of the game.
 * @author Thomas Ranvier
 */
public class Position {
    public static int SIZE = 8;
    private char[][] grid;
    private char player;

    /**
     * Class constructor that takes no arguments.
     */
    public Position() {
        this.grid = new char[this.SIZE][this.SIZE];
    }

    /**
     * Class constructor that takes an argument that represents the state of the game grid.
     * @param pos_str A string representation of the actual grid.
     */
    public Position(String pos_str) {
        this.grid = new char[this.SIZE][this.SIZE];
        this.player = pos_str.charAt(0) == 'W' ? 'O' : 'X';
        this.initializeGrid(pos_str);
    }

    /**
     * Constructs the grid of this class from the string argument.
     * @param pos_str A string representation of the actual grid.
     */
    private void initializeGrid(String pos_str) {
        for (int x = 0; x < this.SIZE; x++) {
            for (int y = 0; y < this.SIZE; y++)
                this.grid[x][y] = pos_str.charAt(1 + x + (y * this.SIZE));
        }
    }

    /**
     * Updates the internal grid of the class depending on the move passed in argument.
     * @param move Move to effectuate.
     */
    public void makeMove(Move move) {
        int x = move.x;
        int y = move.y;
        if (x >= 0 && x < this.SIZE && y >= 0 && y < this.SIZE && !move.is_pass_move) {
            this.grid[x][y] = this.player;
            for (int dir_x = -1; dir_x <= 1; dir_x++) {
                for (int dir_y = -1; dir_y <= 1; dir_y++) {
                    if ((dir_x != 0 || dir_y != 0) && this.checkDirection(x, y, dir_x, dir_y))
                        this.fillDirection(x, y, dir_x, dir_y);
                }
            }
            this.player = this.player == 'X' ? 'O' : 'X';
        }
    }

    /**
     * Changes the cells in the selected direction to the actual player.
     * Stops when it encounters a cell already owned by the player.
     * @param x Coordinate x.
     * @param y Coordinate y.
     * @param dir_x Direction for the axe x.
     * @param dir_y Direction for the axe y.
     */
    private void fillDirection(int x, int y, int dir_x, int dir_y) {
        boolean itself_encountered = false;
        while (!itself_encountered) {
            x += dir_x;
            y += dir_y;
            if (this.grid[x][y] == this.player)
                itself_encountered = true;
            else
                this.grid[x][y] = this.player;
        }
    }

    /**
     * Gives all the possible moves to do for the actual player in the actual game state.
     * @return A list of all the valid moves from the actual state of the game.
     */
    public ArrayList<Move> getValidMoves() {
        ArrayList<Move> valid_moves = new ArrayList<>();
        for (int x = 0; x < this.SIZE; x++) {
            for (int y = 0; y < this.SIZE; y++) {
                if (this.isCandidate(x, y) && this.isValid(x, y))
                    valid_moves.add(new Move(x, y));
            }
        }
        return valid_moves;
    }

    /**
     * Checks if it is valid for the actual player to play in a position in the actual game state.
     * @param x Coordinate x.
     * @param y Coordinate y.
     * @return A boolean indicating if the move is valid or not.
     */
    private boolean isValid(int x, int y) {
        for (int dir_x = -1; dir_x <= 1; dir_x++) {
            for (int dir_y = -1; dir_y <= 1; dir_y++) {
                if ((dir_x != 0 || dir_y != 0) && this.checkDirection(x, y, dir_x, dir_y))
                    return true;
            }
        }
        return false;
    }

    /**
     * Checks if there are only enemies and then a cell owned by the actual player in the selected direction.
     * @param x Coordinate x.
     * @param y Coordinate y.
     * @param dir_x Direction on the axe x.
     * @param dir_y Direction on the axe y.
     * @return A boolean that indicates if the result.
     */
    private boolean checkDirection(int x, int y, int dir_x, int dir_y) {
        boolean out_of_bounds = false;
        boolean enemy_between = false;
        while (!out_of_bounds) {
            x += dir_x;
            y += dir_y;
            if (x < 0 || x >= this.SIZE || y < 0 || y >= this.SIZE) {
                out_of_bounds = true;
            } else {
                if (this.grid[x][y] == this.player)
                    return enemy_between;
                else if (this.grid[x][y] != 'E')
                    enemy_between = true;
                else
                    return false;
            }
        }
        return false;
    }

    /**
     * Checks if the selected cell is a candidate for a valid move.
     * @param x Coordinate x.
     * @param y Coordinate y.
     * @return A boolean true if the cell is empty and has an enemy neighbour and false otherwise.
     */
    private boolean isCandidate(int x, int y) {
        if (this.grid[x][y] != 'E')
            return false;
        return this.hasEnemyNeighbour(x, y);
    }

    /**
     * Checks if the selected cell has an enemy neighbour.
     * @param x Coordinate x.
     * @param y Coordinate y.
     * @return A boolean true if the cell has at least one enemy neighbour, false otherwise.
     */
    private boolean hasEnemyNeighbour(int x, int y) {
        boolean result = false;
        char enemy = this.player == 'X' ? 'O' : 'X';
        if (x > 0 && y > 0 && this.grid[x - 1][y - 1] == enemy)
            result = true;
        if (x > 0 && this.grid[x - 1][y] == enemy)
            result = true;
        if (x > 0 && y < this.SIZE - 1 && this.grid[x - 1][y + 1] == enemy)
            result = true;
        if (y > 0 && this.grid[x][y - 1] == enemy)
            result = true;
        if (y < this.SIZE - 1 && this.grid[x][y + 1] == enemy)
            result = true;
        if (x < this.SIZE - 1 && y > 0 && this.grid[x + 1][y - 1] == enemy)
            result = true;
        if (x < this.SIZE - 1 && this.grid[x + 1][y] == enemy)
            result = true;
        if (x < this.SIZE - 1 && y < this.SIZE - 1 && this.grid[x + 1][y + 1] == enemy)
            result = true;
        return result;
    }

    /**
     * Clones the actual position object.
     * @return A clone of this object.
     */
    public Position clone() {
        Position new_position = new Position();
        new_position.player = this.player;
        for (int x = 0; x < this.SIZE; x++)
            for (int y = 0; y < this.SIZE; y++)
                new_position.grid[x][y] = this.grid[x][y];
        return new_position;
    }

    /**
     * Getter for the player.
     * @return The player.
     */
    public char getPlayer() {
        return this.player;
    }

    /**
     * A getter for the internal grid.
     * @return A clone of the grid.
     */
    public char[][] getGrid() {
        char[][] new_grid = new char[this.SIZE][this.SIZE];
        for (int x = 0; x < this.SIZE; x++) {
            for (int y = 0; y < this.SIZE; y++)
                new_grid[x][y] = this.grid[x][y];
        }
        return new_grid;
    }

    /**
     * Override of the toString function to allow the user to print a simple representation of the internal grid.
     * @return The string that is printed.
     */
    @java.lang.Override
    public java.lang.String toString() {
        String grid_str = " ";
        for (int i = 0; i < this.SIZE - 1; i++)
            grid_str += "--";
        grid_str += "- \n";
        for (int y = 0; y < this.SIZE; y++) {
            grid_str += "|";
            for (int x = 0; x < this.SIZE; x++) {
                grid_str += this.grid[x][y];
                if (x < this.SIZE - 1) {
                    grid_str += " ";
                }
            }
            grid_str += "|\n";
        }
        grid_str += " ";
        for (int i = 0; i < this.SIZE - 1; i++)
            grid_str += "--";
        grid_str += "- ";
        return grid_str;
    }
}